# Loading the Data

In [98]:
import pandas as pd
import numpy as np

In [99]:
features = pd.read_csv("callTextFeatures.csv")
features.head()
features.columns

Index(['Unnamed: 0', '0_Absolute energy_x_x', '0_Area under the curve_x_x',
       '0_Autocorrelation_x_x', '0_Centroid_x_x',
       '0_ECDF Percentile Count_0_x_x', '0_ECDF Percentile Count_1_x_x',
       '0_ECDF Percentile_0_x_x', '0_ECDF Percentile_1_x_x', '0_ECDF_0_x_x',
       ...
       '0_Wavelet variance_0_y_y', '0_Wavelet variance_1_y_y',
       '0_Wavelet variance_2_y_y', '0_Wavelet variance_3_y_y',
       '0_Wavelet variance_4_y_y', '0_Wavelet variance_5_y_y',
       '0_Wavelet variance_6_y_y', '0_Wavelet variance_7_y_y',
       '0_Wavelet variance_8_y_y', '0_Zero crossing rate_y_y'],
      dtype='object', length=602)

In [100]:
Summary = pd.read_csv("SummaryFinal.csv")
Summary.columns

Index(['id', 'Intext', 'OutText', 'InCall', 'InCallDuration', 'OutCall',
       'OutCallDuration', 'appVersion', 'PHQ - Q1', 'PHQ - Q2', 'PHQ - Q3',
       'PHQ - Q4', 'PHQ - Q5', 'PHQ - Q6', 'PHQ - Q7', 'PHQ - Q8', 'PHQ - Q9',
       'PHQ - Total', 'GAD - Q1', 'GAD - Q2', 'GAD - Q3', 'GAD - Q4',
       'GAD - Q5', 'GAD - Q6', 'GAD - Q7', 'GAD - Total', 'Gender', 'Age',
       'StudentStatus', 'PriorDepressionTreatment', 'Remote', 'COVID', 'Group',
       'Timestamp', 'Control', 'Stereotype Threat', '-', 'Man', 'Other',
       'Woman'],
      dtype='object')

I had to go into lab 2 and change the subset of the data to make sure the the features and the summary datasets are
displaying data from the correct subset of data

# Cleaning Up Demographics

In [101]:
Summary = Summary.drop(columns = ['appVersion', 'Age', 'Gender', 'StudentStatus', 'PriorDepressionTreatment', 'Remote', 'COVID', 'Group', 'Timestamp', 'PHQ - Q1', 'PHQ - Q2', 'PHQ - Q3',
       'PHQ - Q4', 'PHQ - Q5', 'PHQ - Q6', 'PHQ - Q7', 'PHQ - Q8', 'PHQ - Q9','GAD - Q1', 'GAD - Q2', 'GAD - Q3', 'GAD - Q4', 'GAD - Q5', 'GAD - Q6', 'GAD - Q7'])

Summary.head()
Summary.columns
#Removed unnecessary demographics

Index(['id', 'Intext', 'OutText', 'InCall', 'InCallDuration', 'OutCall',
       'OutCallDuration', 'PHQ - Total', 'GAD - Total', 'Control',
       'Stereotype Threat', '-', 'Man', 'Other', 'Woman'],
      dtype='object')

In [102]:
Summary.columns = ['id', 'Intext', 'OutText', 'InCall', 'InCallDuration', 'OutCall',
       'OutCallDuration', 'phq9', 'gad7', 'Control',
       'Stereotype Threat', '-', 'Man', 'Other', 'Woman' ]
#Summary.columns
Summary.head()

id  Intext  OutText  InCall  InCallDuration  OutCall  \
0  00WRNS3MK     1.0     64.0    83.0            83.0    181.0   
1  0AS1KO35P    16.0    689.0   244.0           244.0    224.0   
2  0EZE7RBCS    46.0     95.0    98.0            98.0    263.0   
3  0GJVGLPFD     5.0    117.0   211.0           211.0     58.0   
4  0NZH3ULXB   489.0    269.0   247.0           247.0     96.0   

   OutCallDuration  phq9  gad7  Control  Stereotype Threat  -  Man  Other  \
0            181.0    19    17        0                  1  0    1      0   
1            224.0     4     0        1                  0  0    1      0   
2            263.0    25    21        1                  0  0    1      0   
3             58.0     6     1        1                  0  0    1      0   
4             96.0    22    16        1                  0  0    1      0   

   Woman  
0      0  
1      0  
2      0  
3      0  
4      0

# Combining the Datasets on id

In [112]:
#SumandFeat = pd.merge(features, Summary, on='id', how='outer')
#SumandFeat = features.merge(Summary, on = ["id"])
SumandFeat.head(5)

Unnamed: 0  0_Absolute energy_x_x  0_Area under the curve_x_x  \
0           0                   34.0                        0.17   
1           1                   46.0                        0.24   
2           2                 1132.0                        1.54   
3           3                   28.0                        0.16   
4           4                 1188.0                        1.50   

   0_Autocorrelation_x_x  0_Centroid_x_x  0_ECDF Percentile Count_0_x_x  \
0                   34.0        0.082353                            6.0   
1                   46.0        0.115435                            0.0   
2                 1132.0        0.144965                            6.0   
3                   28.0        0.148214                            0.0   
4                 1188.0        0.183687                            0.0   

   0_ECDF Percentile Count_1_x_x  0_ECDF Percentile_0_x_x  \
0                           24.0                      0.0   
1                            0.0                      0.0   
2                           24.0                      2.0   
3                            0.0                      0.0   
4                            0.0                      0.0   

   0_ECDF Percentile_1_x_x  0_ECDF_0_x_x  0_ECDF_1_x_x  0_ECDF_2_x_x  \
0                      1.0      0.032258      0.064516      0.096774   
1                      0.0      0.032258      0.064516      0.096774   
2                      7.0      0.032258      0.064516      0.096774   
3                      0.0      0.032258      0.064516      0.096774   
4                      0.0      0.032258      0.064516      0.096774   

   0_ECDF_3_x_x  0_ECDF_4_x_x  0_ECDF_5_x_x  0_ECDF_6_x_x  0_ECDF_7_x_x  \
0      0.129032       0.16129      0.193548      0.225806      0.258065   
1      0.129032       0.16129      0.193548      0.225806      0.258065   
2      0.129032       0.16129      0.193548      0.225806      0.258065   
3      0.129032       0.16129      0.193548      0.225806      0.258065   
4      0.129032       0.16129      0.193548      0.225806      0.258065   

   0_ECDF_8_x_x  0_ECDF_9_x_x  0_Entropy_x_x  0_FFT mean coefficient_0_x_x  \
0      0.290323      0.322581       0.285989                      0.000245   
1      0.290323      0.322581       0.326658                      0.000047   
2      0.290323      0.322581       0.646669                      0.060782   
3      0.290323      0.322581       0.278260                      0.000629   
4      0.290323      0.322581       0.664606                      0.002219   

   0_FFT mean coefficient_1_x_x  0_FFT mean coefficient_10_x_x  \
0                      0.057882                       0.004231   
1                      0.043406                       0.030247   
2                      0.364986                       0.286213   
3                      0.002579                       0.003884   
4                      0.586811                       0.235620   

   0_FFT mean coefficient_11_x_x  0_FFT mean coefficient_12_x_x  \
0                       0.003460                       0.010484   
1                       0.058202                       0.040966   
2                       0.189471                       0.102293   
3                       0.020472                       0.001893   
4                       0.155319                       0.757985   

   0_FFT mean coefficient_13_x_x  0_FFT mean coefficient_14_x_x  \
0                       0.004765                       0.016727   
1                       0.018862                       0.003633   
2                       0.191317                       0.082932   
3                       0.029149                       0.002472   
4                       0.019511                       0.100806   

   0_FFT mean coefficient_15_x_x  0_FFT mean coefficient_2_x_x  \
0                       0.003119                      0.030812   
1                       0.000480                      0.027487   
2                     

In [111]:
SumandFeat = SumandFeat.fillna(0)
SumandFeat.head(5)

Unnamed: 0  0_Absolute energy_x_x  0_Area under the curve_x_x  \
0           0                   34.0                        0.17   
1           1                   46.0                        0.24   
2           2                 1132.0                        1.54   
3           3                   28.0                        0.16   
4           4                 1188.0                        1.50   

   0_Autocorrelation_x_x  0_Centroid_x_x  0_ECDF Percentile Count_0_x_x  \
0                   34.0        0.082353                            6.0   
1                   46.0        0.115435                            0.0   
2                 1132.0        0.144965                            6.0   
3                   28.0        0.148214                            0.0   
4                 1188.0        0.183687                            0.0   

   0_ECDF Percentile Count_1_x_x  0_ECDF Percentile_0_x_x  \
0                           24.0                      0.0   
1                            0.0                      0.0   
2                           24.0                      2.0   
3                            0.0                      0.0   
4                            0.0                      0.0   

   0_ECDF Percentile_1_x_x  0_ECDF_0_x_x  0_ECDF_1_x_x  0_ECDF_2_x_x  \
0                      1.0      0.032258      0.064516      0.096774   
1                      0.0      0.032258      0.064516      0.096774   
2                      7.0      0.032258      0.064516      0.096774   
3                      0.0      0.032258      0.064516      0.096774   
4                      0.0      0.032258      0.064516      0.096774   

   0_ECDF_3_x_x  0_ECDF_4_x_x  0_ECDF_5_x_x  0_ECDF_6_x_x  0_ECDF_7_x_x  \
0      0.129032       0.16129      0.193548      0.225806      0.258065   
1      0.129032       0.16129      0.193548      0.225806      0.258065   
2      0.129032       0.16129      0.193548      0.225806      0.258065   
3      0.129032       0.16129      0.193548      0.225806      0.258065   
4      0.129032       0.16129      0.193548      0.225806      0.258065   

   0_ECDF_8_x_x  0_ECDF_9_x_x  0_Entropy_x_x  0_FFT mean coefficient_0_x_x  \
0      0.290323      0.322581       0.285989                      0.000245   
1      0.290323      0.322581       0.326658                      0.000047   
2      0.290323      0.322581       0.646669                      0.060782   
3      0.290323      0.322581       0.278260                      0.000629   
4      0.290323      0.322581       0.664606                      0.002219   

   0_FFT mean coefficient_1_x_x  0_FFT mean coefficient_10_x_x  \
0                      0.057882                       0.004231   
1                      0.043406                       0.030247   
2                      0.364986                       0.286213   
3                      0.002579                       0.003884   
4                      0.586811                       0.235620   

   0_FFT mean coefficient_11_x_x  0_FFT mean coefficient_12_x_x  \
0                       0.003460                       0.010484   
1                       0.058202                       0.040966   
2                       0.189471                       0.102293   
3                       0.020472                       0.001893   
4                       0.155319                       0.757985   

   0_FFT mean coefficient_13_x_x  0_FFT mean coefficient_14_x_x  \
0                       0.004765                       0.016727   
1                       0.018862                       0.003633   
2                       0.191317                       0.082932   
3                       0.029149                       0.002472   
4                       0.019511                       0.100806   

   0_FFT mean coefficient_15_x_x  0_FFT mean coefficient_2_x_x  \
0                       0.003119                      0.030812   
1                       0.000480                      0.027487   
2                     

# Making Phq9 and Gad7 Binary

In [113]:
split = 10
data = SumandFeat
d10 = []
g10 = []
for i in range(0, data.shape[0]):
    if int(data.phq9[i]) >= split:
        d10.append(1)
    else:
        d10.append(0)
    if int(data.gad7[i]) >= split:
        g10.append(1)
    else:
        g10.append(0)
data["phq9"] = d10
data["gad7"] = g10

data.head()


Unnamed: 0  0_Absolute energy_x_x  0_Area under the curve_x_x  \
0           0                   34.0                        0.17   
1           1                   46.0                        0.24   
2           2                 1132.0                        1.54   
3           3                   28.0                        0.16   
4           4                 1188.0                        1.50   

   0_Autocorrelation_x_x  0_Centroid_x_x  0_ECDF Percentile Count_0_x_x  \
0                   34.0        0.082353                            6.0   
1                   46.0        0.115435                            0.0   
2                 1132.0        0.144965                            6.0   
3                   28.0        0.148214                            0.0   
4                 1188.0        0.183687                            0.0   

   0_ECDF Percentile Count_1_x_x  0_ECDF Percentile_0_x_x  \
0                           24.0                      0.0   
1                            0.0                      0.0   
2                           24.0                      2.0   
3                            0.0                      0.0   
4                            0.0                      0.0   

   0_ECDF Percentile_1_x_x  0_ECDF_0_x_x  0_ECDF_1_x_x  0_ECDF_2_x_x  \
0                      1.0      0.032258      0.064516      0.096774   
1                      0.0      0.032258      0.064516      0.096774   
2                      7.0      0.032258      0.064516      0.096774   
3                      0.0      0.032258      0.064516      0.096774   
4                      0.0      0.032258      0.064516      0.096774   

   0_ECDF_3_x_x  0_ECDF_4_x_x  0_ECDF_5_x_x  0_ECDF_6_x_x  0_ECDF_7_x_x  \
0      0.129032       0.16129      0.193548      0.225806      0.258065   
1      0.129032       0.16129      0.193548      0.225806      0.258065   
2      0.129032       0.16129      0.193548      0.225806      0.258065   
3      0.129032       0.16129      0.193548      0.225806      0.258065   
4      0.129032       0.16129      0.193548      0.225806      0.258065   

   0_ECDF_8_x_x  0_ECDF_9_x_x  0_Entropy_x_x  0_FFT mean coefficient_0_x_x  \
0      0.290323      0.322581       0.285989                      0.000245   
1      0.290323      0.322581       0.326658                      0.000047   
2      0.290323      0.322581       0.646669                      0.060782   
3      0.290323      0.322581       0.278260                      0.000629   
4      0.290323      0.322581       0.664606                      0.002219   

   0_FFT mean coefficient_1_x_x  0_FFT mean coefficient_10_x_x  \
0                      0.057882                       0.004231   
1                      0.043406                       0.030247   
2                      0.364986                       0.286213   
3                      0.002579                       0.003884   
4                      0.586811                       0.235620   

   0_FFT mean coefficient_11_x_x  0_FFT mean coefficient_12_x_x  \
0                       0.003460                       0.010484   
1                       0.058202                       0.040966   
2                       0.189471                       0.102293   
3                       0.020472                       0.001893   
4                       0.155319                       0.757985   

   0_FFT mean coefficient_13_x_x  0_FFT mean coefficient_14_x_x  \
0                       0.004765                       0.016727   
1                       0.018862                       0.003633   
2                       0.191317                       0.082932   
3                       0.029149                       0.002472   
4                       0.019511                       0.100806   

   0_FFT mean coefficient_15_x_x  0_FFT mean coefficient_2_x_x  \
0                       0.003119                      0.030812   
1                       0.000480                      0.027487   
2                     

In [114]:
#checking values
print(sum(d10)/len(d10))
print(len(d10))
print(sum(g10)/len(g10))
print(len(g10))

0.5435356200527705
379
0.41952506596306066
379


# Running the Machine Learning Code

In [115]:
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.decomposition import PCA
from sklearn.utils import resample
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.metrics import confusion_matrix

In [116]:
numberOfFeatures = 2 #set number of principal components
modelTypelist = ["SVC", "kNN", "RF", "LR", "XG"]

ba_depression = []
ba_anxiety = []
rlist = []
mlist = []

for modelType in modelTypelist: 
    for r in range(50, 75):
        print(r) 
        rlist.append(r)
        mlist.append(modelType)

        #create train/test sets
        df_train, df_test = train_test_split(data, test_size=0.3, stratify=data[["phq9", "gad7"]], random_state = r)

        #save target variables
        train_targets = df_train.loc[:,["phq9","gad7"]]
        test_targets = df_test.loc[:,["phq9","gad7"]]

        #drop non-features: id and targets
        trainContent = df_train.drop(columns = ["id", "phq9", "gad7"])
        testContent = df_test.drop(columns = ["id", "phq9", "gad7"])

        #normalize before pca
        min_max_scaler = preprocessing.MinMaxScaler()  
        np_scaled = min_max_scaler.fit_transform(trainContent)
        featureSubset = pd.DataFrame(np_scaled)
        np_scaled2 =  min_max_scaler.transform(testContent)
        testSubset = pd.DataFrame(np_scaled2)

        #principal component analysis
        pca = PCA(n_components=numberOfFeatures)
        pca = pca.fit(featureSubset)
        X_pca = pca.transform(featureSubset)
        pcaDF = pd.DataFrame(X_pca)
        testSubset2 = pca.transform(testSubset)
        pca_test = pd.DataFrame(testSubset2)

        #upsampling to balance classes
        train_targets = train_targets.reset_index(drop = True)
        pcaDF_labels = pd.concat([pcaDF,train_targets], axis = 1) #reattach labels

        phq0 = pcaDF_labels[pcaDF_labels.phq9 == 0]
        phq1 = pcaDF_labels[pcaDF_labels.phq9 == 1]
        if phq0.shape[0]>phq1.shape[0]:
            phq_upsampled = resample(phq1, n_samples=(phq0.shape[0]-phq1.shape[0]), random_state=50)
            pca_phq_up = pd.concat([pcaDF_labels, phq_upsampled])
        elif phq0.shape[0]<phq1.shape[0]:
            phq_upsampled = resample(phq0, n_samples=(phq1.shape[0]-phq0.shape[0]), random_state=50)
            pca_phq_up = pd.concat([pcaDF_labels, phq_upsampled])

        gad0 = pcaDF_labels[pcaDF_labels.gad7 == 0]
        gad1 = pcaDF_labels[pcaDF_labels.gad7 == 1]

        if gad0.shape[0]>gad1.shape[0]:
            gad_upsampled = resample(gad1, n_samples=(gad0.shape[0]-gad1.shape[0]), random_state=50)
            pca_gad_up = pd.concat([pcaDF_labels, gad_upsampled])
        elif gad0.shape[0]<gad1.shape[0]:
            gad_upsampled = resample(gad0, n_samples=(gad1.shape[0]-gad0.shape[0]), random_state=50)
            pca_gad_up = pd.concat([pcaDF_labels, gad_upsampled])

        #remove upsampled labels
        pca_phq_targets = pca_phq_up.loc[:,["phq9","gad7"]]
        pca_phq_features = pca_phq_up.drop(columns = ["phq9", "gad7"])

        pca_gad_targets = pca_gad_up.loc[:,["phq9","gad7"]]
        pca_gad_features = pca_gad_up.drop(columns = ["phq9", "gad7"])

        #select modeltype
        if modelType == "SVC":
            clf = svm.SVC(random_state=r)
        elif modelType == "RF":
            clf = RandomForestClassifier(random_state=r)
        elif modelType == "kNN":
            clf = KNeighborsClassifier()
        elif modelType == "LR":
            clf = LogisticRegression(random_state=r)
        
        for label in ["phq9","gad7"]:
            #fit model and make predictions
            clf.fit(pca_phq_features, list(pca_phq_targets[label]))
            y_pred = clf.predict(pca_test)

            #evaluate model
            conf_mat = confusion_matrix(list(test_targets[label]), y_pred)
            TN = conf_mat[0][0]
            TP = conf_mat[1][1]
            FP = conf_mat[0][1]
            FN = conf_mat[1][0]
            sens = TP/(TP+FN)
            spec = TN/(TN+FP)
            ba = (sens + spec)/2
            
            if label == "phq9":
                ba_depression.append(ba)
            elif label == "gad7":
                ba_anxiety.append(ba)
                
resultsDF = pd.DataFrame()
resultsDF["ba_deppression"] = ba_depression
resultsDF["ba_anxiety"] = ba_anxiety
resultsDF["random"] = rlist
resultsDF["model"] = mlist

50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74


In [132]:

resultsDF.to_csv("Lab4Results.csv")
resultsDF.head(2)

ba_deppression  ba_anxiety  random model
0        0.526675         0.5      50   SVC
1        0.504963         0.5      51   SVC

# Exploring the Results

In [123]:
import plotly.express as px
fig = px.box(resultsDF, x="model", y="ba_deppression")
fig.show()

In [129]:
fig1 = px.box(resultsDF, x="model", y="ba_anxiety")
fig1.show()

Changed subset based on model, what here is the last model subset

In [142]:
XG = resultsDF[resultsDF.model == "XG"].reset_index()
XG.head(5)
print(sum(XG.ba_deppression)/len(XG.ba_deppression))
print(sum(XG.ba_anxiety)/len(XG.ba_anxiety))

0.5430769230769232
0.5054545454545454


In [145]:
XGstda = np.std(XG.ba_anxiety)
XGstda

0.012263253919503366

# Questions

1. I used data after June 1, 2020 becuase I originally had after April 2021 and the dataset was too small. I used the direction as 1 for incoming calls and texts and direction as 2 for outgoing calls and texts. The subset is the same for thqe count logs and the time series.

2. For ba_Depression The averages for each are similar and LR and XG are exactly the same,
    For ba_anxiety there are more differences in the boxplots some are small RF has the widest range but has outliers

3. SVC Depression = 0.526, Anxiety = 0.497,   kNN Depression = 0.513 Anxiety = 0.491,  RF Depression = 0.502 Anxiety = 0.493,  LR Depression = 0.543 Anxiety = 0.505,  XG Depression = 0.543 Anxiety = 0.505,  The XG and LR models are the highest which happen to be the same

4. XG Standard Deviation,   Depression = 0.041 Anxiety = 0.012

5. XG Depression Highest = 0.631 Lowest = 0.466 Anxiety Highest = 0.542 Lowest = 0.488. random_state is used to make the sampling fo the data random.

6. Drops:trainContent = df_train.drop(columns = ["id", "phq9", "gad7"]), Subsets: phq1 = pcaDF_labels[pcaDF_labels.phq9 == 1]
   Selects: if modelType == "SVC": clf = svm.SVC(random_state=r)

7. Because you're concatenating labels in the next line so have to make sure the index is in order so that this works properly 

8. Resampling is used to take different samples of the same data with and without replacement. n_samples keeps track of the amount of samples 